script/heatmap.py

```diff
@@ -0,0 +1,277 @@
+            if self.toPos[0] > line_no:
+                # the region spans the whole line
+                return (1, width)
+            else:
+                # the region starts on a preceding line but end on this one
+                return (1, self.toPos[1])
+        else:
+            if self.toPos[0] > line_no:
+                # the region starts on this line but ends later
+                return (self.fromPos[1], width)
+            else:
+                # the region starts and ends on this line
+                return (self.fromPos[1], self.toPos[1])
+
+    def isInside(self, other):
+        self.filename = other.filename \
+                and self.fromPos >= outer.fromPos \
+                and self.toPos <= outer.toPos
+
+
+# the regular expression for the source location as produced by SANY
+REGION_RE = re.compile(r"(.*):(\d+):(\d+)-(\d+):(\d+)")
+
+def region_from_csv(row):
+    m = REGION_RE.match(row[4])
+    if m:
+        entry = ProfileEntry(int(row[0]), int(row[1]), int(row[2]), int(row[3]))
+        return CodeRegion(m.group(1),
+                      (int(m.group(2)), int(m.group(3))), 
+                      (int(m.group(4)), int(m.group(5))),
+                      entry)
+    else:
+        print("Invalid source location: {}".format(row[4]), file=sys.stderr)
+        return None
+
+
+def merge_same_regions(sorted_regions):
+    """The profiling data contains several points for the same range. Merge them."""
+    prev_r = None
+    new_regions = []
+    for r in sorted_regions:
+        if prev_r and prev_r.fromPos == r.fromPos and prev_r.toPos == r.toPos:
+            # merge the data points
+            prev_r.profEntry.add(r.profEntry)
+        else:
+            new_regions.append(r)
+
+        prev_r = r
+
+    return new_regions
+
+
+def heat_color(weight, max_weight):
+    h = int(110 * (1.0 - float(weight) / max_weight))
+    return "hsl({}, 100%, 50%)".format(h)
+
+
+def find_max_weight(regions):
+    """Find the maximal weight in the given regions"""
+
+    mw = 0
+    for r in regions:
+        mw = max(mw, r.profEntry.weight)
+
+    return mw
+
+
+def write_heatmap(heatfile, code_filename, regions):
+    """Write code lines with while coloring them by code regions"""
+
+
+    def drop_old_regions(regions, line_no):
+        # remove the regions that end before line_no
+        return list(itertools.dropwhile(lambda r: r.toPos[0] < line_no, regions))
+
+    def split_line_by_regions(regions_ahead, line, line_no):
+        line_len = len(line)
+        # take the regions that cover the current line
+        window = itertools.takewhile(lambda r: r.fromPos[0] <= line_no, regions_ahead)
+        line_regions = sorted(list(window),
+                # first, sort by the starting column; then by the nr. of lines
+                key=lambda r: \
+                        (r.rangeByLine(line_no, line_len)[0], r.toPos[0] - r.fromPos[0]))
+
+        #print("Line: {}, length: {}".format(line_no, line_len))
```

Comment:
did you mean to leave these in for debugging purposes? (also on line 143 and 146)

Reply:
Yes. These were quite useful. If there is a problem, we will need something similar.